src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,586 @@
+
+  EXPECT_EQ(done_status, Status::OK);
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cache: true, Callback: in place
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestCachedQuotaWithInPlaceCallbackInSyncMultipleRequests) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillOnce(Invoke(&mock_quota_transport_,
+                       &MockQuotaTransport::AllocateQuotaWithInplaceCallback));
+
+  AllocateQuotaResponse quota_response;
+
+  for (int i = 0; i < 10; i++) {
+    Status done_status = client_->Quota(quota_request1_, &quota_response);
+    EXPECT_EQ(done_status, Status::OK);
+  }
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 10);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cache: false, Callback: thread
+TEST_F(ServiceControlClientImplQuotaTest, TestNonCachedQuotaThread) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillRepeatedly(Invoke(&mock_quota_transport_,
+                             &MockQuotaTransport::AllocateQuotaUsingThread));
+
+  Status done_status;
+  AllocateQuotaResponse quota_response;
+
+  // Set the check status and response to be used in the on_check_done
+  mock_quota_transport_.done_status_ = done_status;
+  mock_quota_transport_.quota_response_ = &quota_response;
+
+  StatusPromise status_promise;
+  StatusFuture status_future = status_promise.get_future();
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&status_promise, &done_status](Status status) {
+                   StatusPromise moved_promise(std::move(status_promise));
+                   moved_promise.set_value(status);
+                 });
+
+  // Since it is not cached, transport should be called.
+  EXPECT_TRUE(MessageDifferencer::Equals(mock_quota_transport_.quota_request_,
+                                         quota_request1_));
+
+  status_future.wait();
+  EXPECT_EQ(status_future.get(), Status::OK);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cache: true, Callback: thread
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestCachedQuotaInSyncThreadMultipleRequests) {
```

Comment:
This is not "InSync".   "InSync" mean calling the way as line 487, status is returned by the function, not from the callback. 

I suggest keep this case, and add another case for "InSync"